File: frappe_app/facility_os/facility_os/api/validation.py
import frappe
from frappe.utils import now_datetime
from facility_os.api.security import (
    ROLE_ADMIN,
    ROLE_INVENTORY,
    current_user,
    require_any_role,
)

TRANSITIONS = {
    "submit": ("Draft", "Submitted", ROLE_INVENTORY),
    "inventory_validate": ("Submitted", "Inventory Validated", ROLE_INVENTORY),
    "inventory_reject": ("Submitted", "Inventory Rejected", ROLE_INVENTORY),
    "admin_approve": ("Inventory Validated", "Admin Approved", ROLE_ADMIN),
    "admin_reject": ("Inventory Validated", "Admin Rejected", ROLE_ADMIN),
    "reopen": (None, "Draft", ROLE_INVENTORY),
}

REJECT_ACTIONS = {"inventory_reject", "admin_reject"}


def _audit(doc, action, previous, new_status, actor, actor_role, remarks):
    frappe.get_doc({
        "doctype": "Facility Validation Record",
        "reference_doctype": doc.doctype,
        "reference_name": doc.name,
        "action": action,
        "previous_status": previous,
        "new_status": new_status,
        "actor": actor,
        "actor_role": actor_role,
        "remarks": remarks or "",
        "blocking_exceptions": doc.blocking_exceptions or 0,
        "occurred_at": now_datetime(),
    }).insert(ignore_permissions=True)


@frappe.whitelist(methods=["POST"])
def act(session_name, action, remarks=None):
    if action not in TRANSITIONS:
        frappe.throw("Unsupported validation action.")

    expected, new_status, required_role = TRANSITIONS[action]
    require_any_role(required_role)
    actor = current_user()
    doc = frappe.get_doc("Facility Physical Count Session", session_name)
    previous = doc.validation_status or "Draft"

    if action == "reopen":
        if previous not in ("Inventory Rejected", "Admin Rejected"):
            frappe.throw("Only a rejected validation can be reopened.")
    elif previous != expected:
        frappe.throw(f"Action {action} is not allowed from status {previous}.")

    if action in REJECT_ACTIONS and not (remarks or "").strip():
        frappe.throw("Remarks are required when rejecting validation.")

    if action in ("inventory_validate", "admin_approve") and int(doc.blocking_exceptions or 0) > 0:
        frappe.throw("Blocking reconciliation exceptions must be resolved first.")

    if action == "admin_approve" and doc.inventory_validated_by == actor:
        frappe.throw("Admin approver must be different from the Inventory validator.")

    if action == "inventory_validate":
        doc.inventory_validated_by = actor
        doc.inventory_validated_at = now_datetime()
    elif action == "admin_approve":
        doc.admin_approved_by = actor
        doc.admin_approved_at = now_datetime()
        doc.opening_stock_gate = "Ready for Approval"
    else:
        doc.opening_stock_gate = "Blocked"

    doc.validation_status = new_status
    doc.save(ignore_permissions=True)
    _audit(doc, action, previous, new_status, actor, required_role, remarks)
    frappe.db.commit()

    return {
        "ok": True,
        "currentStatus": previous.upper().replace(" ", "_"),
        "nextStatus": new_status.upper().replace(" ", "_"),
        "openingStockGate": "READY_FOR_APPROVAL" if doc.opening_stock_gate == "Ready for Approval" else "BLOCKED",
        "actor": actor,
        "persisted": True,
        "message": "Validation recorded.",
    }

File: lib/reconciliation/validation.ts
export type ValidationRole = 'inventory' | 'admin';

export type ValidationStatus =
  | 'DRAFT'
  | 'SUBMITTED'
  | 'INVENTORY_VALIDATED'
  | 'INVENTORY_REJECTED'
  | 'ADMIN_APPROVED'
  | 'ADMIN_REJECTED';

export type ValidationAction =
  | 'submit'
  | 'inventory_validate'
  | 'inventory_reject'
  | 'admin_approve'
  | 'admin_reject'
  | 'reopen';

export type ValidationRequest = {
  action: ValidationAction;
  currentStatus: ValidationStatus;
  actorRole: ValidationRole;
  actorName?: string;
  remarks?: string;
  blockingExceptions?: number;
};

export type ValidationPreview = {
  ok: boolean;
  currentStatus: ValidationStatus;
  nextStatus: ValidationStatus;
  openingStockGate: 'BLOCKED' | 'READY_FOR_APPROVAL';
  message: string;
  errors: string[];
  auditEvent?: {
    action: ValidationAction;
    actorRole: ValidationRole;
    actorName: string;
    remarks: string;
    occurredAt: string;
  };
};

function allowed(
  action: ValidationAction,
  status: ValidationStatus,
  role: ValidationRole,
) {
  if (action === 'submit') return status === 'DRAFT' && role === 'inventory';
  if (action === 'inventory_validate') return status === 'SUBMITTED' && role === 'inventory';
  if (action === 'inventory_reject') return status === 'SUBMITTED' && role === 'inventory';
  if (action === 'admin_approve') return status === 'INVENTORY_VALIDATED' && role === 'admin';
  if (action === 'admin_reject') return status === 'INVENTORY_VALIDATED' && role === 'admin';
  if (action === 'reopen') return (status === 'INVENTORY_REJECTED' || status === 'ADMIN_REJECTED') && role === 'inventory';
  return false;
}

function nextStatus(action: ValidationAction): ValidationStatus {
  if (action === 'submit') return 'SUBMITTED';
  if (action === 'inventory_validate') return 'INVENTORY_VALIDATED';
  if (action === 'inventory_reject') return 'INVENTORY_REJECTED';
  if (action === 'admin_approve') return 'ADMIN_APPROVED';
  if (action === 'admin_reject') return 'ADMIN_REJECTED';
  return 'DRAFT';
}

export function previewValidation(input: ValidationRequest): ValidationPreview {
  const errors: string[] = [];
  const blockingExceptions = Number(input.blockingExceptions ?? 0);

  if (!allowed(input.action, input.currentStatus, input.actorRole)) {
    errors.push('This role/action is not allowed from the current validation status.');
  }

  if ((input.action === 'inventory_reject' || input.action === 'admin_reject') && !input.remarks?.trim()) {
    errors.push('Remarks are required when rejecting a validation.');
  }

  if (input.action === 'inventory_validate' && blockingExceptions > 0) {
    errors.push('Inventory validation cannot pass while blocking reconciliation exceptions remain.');
  }

  if (input.action === 'admin_approve' && blockingExceptions > 0) {
    errors.push('Admin approval cannot pass while blocking reconciliation exceptions remain.');
  }

  const proposed = nextStatus(input.action);
  const ok = errors.length === 0;
  const finalStatus = ok ? proposed : input.currentStatus;

  return {
    ok,
    currentStatus: input.currentStatus,
    nextStatus: finalStatus,
    openingStockGate: finalStatus === 'ADMIN_APPROVED' && blockingExceptions === 0
      ? 'READY_FOR_APPROVAL'
      : 'BLOCKED',
    message: ok
      ? finalStatus === 'ADMIN_APPROVED'
        ? 'Admin validation approved. Opening stock is eligible for a separate posting approval step.'
        : `Validation moved to ${finalStatus}.`
      : 'Validation preview failed.',
    errors,
    auditEvent: ok ? {
      action: input.action,
      actorRole: input.actorRole,
      actorName: input.actorName?.trim() || 'Current user',
      remarks: input.remarks?.trim() || '',
      occurredAt: new Date().toISOString(),
    } : undefined,
  };
}
